cluster-autoscaler/simulator/delta_cluster_snapshot.go

```diff
@@ -0,0 +1,484 @@
+	}
+	data.addedNodeInfoMap[nodeInfo.Node().Name] = nodeInfo
+	if data.nodeInfoList != nil {
+		data.nodeInfoList = append(data.nodeInfoList, nodeInfo)
+	}
+	return nil
+}
+
+func (data *internalDeltaSnapshotData) clearCaches() {
+	data.nodeInfoList = nil
+	data.clearPodCaches()
+}
+
+func (data *internalDeltaSnapshotData) clearPodCaches() {
+	data.podList = nil
+	data.havePodsWithAffinity = nil
+}
+
+func (data *internalDeltaSnapshotData) updateNode(node *schedulernodeinfo.NodeInfo) error {
+	if _, found := data.addedNodeInfoMap[node.Node().Name]; found {
+		data.removeNode(node.Node().Name)
+	}
+	if _, found := data.modifiedNodeInfoMap[node.Node().Name]; found {
+		data.removeNode(node.Node().Name)
+	}
+
+	return data.addNodeInfo(node)
+}
+
+func (data *internalDeltaSnapshotData) removeNode(nodeName string) error {
+	_, foundInDelta := data.addedNodeInfoMap[nodeName]
+	if foundInDelta {
+		// If node was added within this delta, delete this change.
+		delete(data.addedNodeInfoMap, nodeName)
+	}
+
+	if _, modified := data.modifiedNodeInfoMap[nodeName]; modified {
+		// If node was modified within this delta, delete this change.
+		delete(data.modifiedNodeInfoMap, nodeName)
+	}
+
+	_, foundInBase := data.baseData.getNodeInfo(nodeName)
+	if foundInBase {
+		// If node was found in the underlying data, mark it as deleted in delta.
+		data.deletedNodeInfos[nodeName] = true
+	}
+
+	if !foundInBase && !foundInDelta {
+		// Node not found in the chain.
+		return errNodeNotFound
+	}
+
+	// Maybe consider deleting from the lists instead. Maybe not.
+	data.clearCaches()
+	return nil
+}
+
+func (data *internalDeltaSnapshotData) addPod(pod *apiv1.Pod, nodeName string) error {
+	dni, found := data.getNodeInfoLocal(nodeName)
+	if !found {
+		bni, found := data.baseData.getNodeInfo(nodeName)
+		if !found {
+			return errNodeNotFound
+		}
+		dni = bni.Clone()
+		data.modifiedNodeInfoMap[nodeName] = dni
+		data.clearCaches()
+	}
+
+	dni.AddPod(pod)
+	if data.podList != nil || data.havePodsWithAffinity != nil {
+		data.clearPodCaches()
+	}
+	return nil
+}
+
+func (data *internalDeltaSnapshotData) removePod(namespace, name, nodeName string) error {
+	dni, found := data.getNodeInfoLocal(nodeName)
+	if !found {
+		bni, found := data.baseData.getNodeInfo(nodeName)
+		if !found {
+			return errNodeNotFound
+		}
+		dni = bni.Clone()
+		data.modifiedNodeInfoMap[nodeName] = dni
+		data.clearCaches()
+	}
+
+	preAffinityPods := len(dni.PodsWithAffinity())
+
+	podFound := false
+	for _, pod := range dni.Pods() {
+		if pod.Namespace == namespace && pod.Name == name {
+			if err := dni.RemovePod(pod); err != nil {
+				return fmt.Errorf("cannot remove pod; %v", err)
+			}
+			podFound = true
+			break
+		}
+	}
+	if !podFound {
+		return fmt.Errorf("pod %s/%s not in snapshot", namespace, name)
+	}
+
+	// Maybe consider deleting from the list in the future. Maybe not.
+	postAffinityPods := len(dni.PodsWithAffinity())
+	if preAffinityPods == 1 && postAffinityPods == 0 {
```

Comment:
I wonder if we need this logic at all. I question how likely we're to call removePod() while having cached havePodsWithAffinity, but not podList.